sdk/webpubsub/Microsoft.Azure.WebJobs.Extensions.WebPubSub/src/Utilities.cs

```diff
@@ -77,80 +76,89 @@ public static HttpResponseMessage BuildConnectEventResponse(string response, Dic
         {
             HttpResponseMessage result = new();
 
-            result.StatusCode = GetStatusCode(error.Code);
-            result.Content = new StringContent(error.ErrorMessage);
+            result.StatusCode = GetStatusCode(code);
+            result.Content = new StringContent(errorMessage);
             return result;
         }
 
         public static HttpResponseMessage BuildValidResponse(
-            object response, RequestType requestType,
+            WebPubSubEventResponse response, RequestType requestType,
             WebPubSubConnectionContext context)
         {
-            JsonDocument converted = null;
-            string originStr = null;
-            bool needConvert = true;
-            if (response is WebPubSubEventResponse)
+            // check error as top priority.
+            if (response is EventErrorResponse errorResponse)
             {
-                needConvert = false;
+                return BuildErrorResponse(errorResponse);
+            }
+
+            if (requestType == RequestType.Connect)
+            {
+                if (response is ConnectEventResponse connectResponse)
+                {
+                    var mergedStates = context.UpdateStates(connectResponse.States);
+                    return BuildConnectEventResponse(JsonConvert.SerializeObject(response), mergedStates);
+                }
+                return BuildErrorResponse($"Invalid response type: '{response.GetType()}' in current request type '{requestType}'.");
             }
-            else
+            if (requestType == RequestType.User)
             {
-                // JObject or string type, use string to convert between JObject and JsonDocument.
-                originStr = response.ToString();
-                converted = JsonDocument.Parse(originStr);
+                if (response is UserEventResponse messageResponse)
+                {
+                    var mergedStates = context.UpdateStates(messageResponse.States);
+                    return BuildUserEventResponse(messageResponse, mergedStates);
+                }
+                return BuildErrorResponse($"Invalid response type: '{response.GetType()}' in current request type '{requestType}'.");
             }
+            // should not hit.
+            throw new Exception($"Invalid request type, {requestType}");
+        }
 
+        public static HttpResponseMessage BuildValidResponse(
+            JToken jResponse, RequestType requestType,
+            WebPubSubConnectionContext context)
+        {
             try
             {
-                // Check error, errorCode is required for json convert, otherwise, ignored.
-                if (needConvert && converted.RootElement.TryGetProperty("code", out var code))
+                // Accepts JObject response or stringified, otherwise return ServerError.
+                JObject response = jResponse is JObject jObj ? jObj : JObject.Parse(jResponse.ToString());
```

Comment:
in which case is a `JToken.ToString` is a `JObject` while `JToken` itself is not a `JObject`? If you want to support a string JToken, check if JToken is a string instead of using `ToString`?

Reply:
A corner case that user might return stringified response and load as JValue, and ToString() will return the JObject json value. Tried to allow some corner cases in type-less language, or maybe throw exceptions to strict the input is safer.